### ISSUE TITLE
fuse-archive should skip entries containing a redundant hardlink

### ISSUE THREAD

**Issue by og900aero** [google/fuse-archive#27]
I use Debian 12 with Bspwm tiling windows manager. I installed fuse-archive, and try mount my archive, but I get error message:
fuse-archive: duplicate pathname in /home/Data/Linux/Backup/valami.tar.zst: /.gtkrc-2.0

This archive contains full of home directory.

When I use archivemount for mounting archive, then works normally.

**Comment by fdegros** [comment]
Thanks for the report.

Please try `fuse-archive` version 1.0, which I recently released. I think it will fix your issue. Let me know how it is going.

**Comment by fdegros** [comment]
Thanks for the confirmation.

Going back to the root of the issue, it looks like your archive contains a duplicated entry. Older versions of `fuse-archive` simply rejected archives with duplicated entries, but the newest version 1.0 now gracefully deduplicates the colliding entries by adding a number to their filename, which might or might not be what you expect.

You can confirm this is happening by looking at the verbose logs. Example:

```shell
fuse-archive --verbose /home/Data/Linux/Backup/valami.tar.zst mnt |& grep conflict
```

Now, I'm curious to understand how you ended up with an archive containing a duplicated entry. Were you aware there was a duplicated entry in your archive? Was it done on purpose, or was it an accident? If it was on purpose, what do you expect to see in the mounted archive?

**Comment by og900aero** [comment]
> Thanks for the confirmation.
> 
> Going back to the root of the issue, it looks like your archive contains a duplicated entry. Older versions of `fuse-archive` simply rejected archives with duplicated entries, but the newest version 1.0 now gracefully deduplicates the colliding entries by adding a number to their filename, which might or might not be what you expect.
> 
> You can confirm this is happening by looking at the verbose logs. Example:
> 
> fuse-archive --verbose /home/Data/Linux/Backup/valami.tar.zst mnt |& grep conflict
> Now, I'm curious to understand how you ended up with an archive containing a duplicated entry. Were you aware there was a duplicated entry in your archive? Was it done on purpose, or was it an accident? If it was on purpose, what do you expect to see in the mounted archive?

It is very intresting. When I run with verbose options, I get the following rows:
```
fuse-archive: File [7,962] '/.gtkrc-2.0' conflicts with File [7,962] '/.gtkrc-2.0'
fuse-archive: Resolved conflict for File [7,962] '/.gtkrc-2 (1).0'
fuse-archive: Resolved Hardlink '/.gtkrc-2 (1).0' -> File [7,962] '/.gtkrc-2.0'
```

But when I exctract this archive, I get only one .gtkrc-2.0 file not two.

**Comment by fdegros** [comment]
That's odd. I am not expecting a file entry to conflict with itself. I would like to understand what is really happening here.

When you look at the mounted archive, does the file named `.gtkrc-2.0` have the expected contents and attributes? Can you also see a file named `.gtkrc-2 (1).0`? Does it have the same attributes and contents as `.gtkrc-2.0`?

When you list the contents of your archive with `tar tvf`, how many entries do you see for `.gtkrc-2.0`?

```
tar tvf /home/Data/Linux/Backup/valami.tar.zst | grep gtkrc
```

**Comment by fdegros** [comment]
Can you please build and try `fuse-archive` at commit ad4ae3ed685496ba897ca4a373a777212e6a7152? I believe this should silently fix your issue without introducing any redundant file entries. Let me know how it is going.

**Comment by og900aero** [comment]
> When you look at the mounted archive, does the file named `.gtkrc-2.0` have the expected contents and attributes?

When fuse-archive mounted to the directory, I see the following:
```
.rw-r--r--   599 root root  28 May  2023  '.gtkrc-2 (1).0'
.rw-r--r--   599 root root  28 May  2023  .gtkrc-2.0
```

When I use stat command (seem to be fine):
```
╭(lf)─   shyciii@DEBIAN  /mnt/share  took 14ms 
╰────$ stat .gtkrc-2.0 
  File: .gtkrc-2.0
  Size: 599             Blocks: 2          IO Block: 512    regular file
Device: 0,46    Inode: 7965        Links: 2
Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2024-12-03 10:29:55.000000000 +0100
Modify: 2023-05-28 16:51:45.000000000 +0200
Change: 2024-12-03 10:29:55.000000000 +0100
 Birth: -

╭(lf)─   shyciii@DEBIAN  /mnt/share  took 13ms 
╰────$ stat '.gtkrc-2 (1).0' 
  File: .gtkrc-2 (1).0
  Size: 599             Blocks: 2          IO Block: 512    regular file
Device: 0,46    Inode: 7965        Links: 2
Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2024-12-03 10:29:55.000000000 +0100
Modify: 2023-05-28 16:51:45.000000000 +0200
Change: 2024-12-03 10:29:55.000000000 +0100
 Birth: -
```
When I use tar tvf /home/Data/Linux/Backup/valami.tar.zst | grep gtkrc ,then I get this strange:

```
-rw-r--r-- shyciii/users        599 2023-05-28 16:51 .gtkrc-2.0
hrw-r--r-- shyciii/users          0 2023-05-28 16:51 .gtkrc-2.0 link to .gtkrc-2.0

```

**Comment by og900aero** [comment]
> Can you please build and try `fuse-archive` at commit [ad4ae3e](https://github.com/google/fuse-archive/commit/ad4ae3ed685496ba897ca4a373a777212e6a7152)? I believe this should silently fix your issue without introducing any redundant file entries. Let me know how it is going.

How can I compile it with this commit? Or just git clone again and make install?

**Comment by fdegros** [comment]
Thanks for the verification.

> How can I compile it with this commit? Or just git clone again and make install?

Yes, you can just build it with the latest sources.

> ```
> -rw-r--r-- shyciii/users        599 2023-05-28 16:51 .gtkrc-2.0
> hrw-r--r-- shyciii/users          0 2023-05-28 16:51 .gtkrc-2.0 link to .gtkrc-2.0
> ```

Here it is. Your archive contains an entry which is a hard link... to itself.

I had never seen that before. Now that I know it can happen, I modified `fuse-archive` to simply ignore (i.e. skip) these redundant entries. Example:

```
fuse-archive --verbose /home/Data/Linux/Backup/valami.tar.zst mnt |& grep Skip
```

You should see the redundant link for `/.gtkrc-2.0` getting skipped.

Your archive might also contain other redundant links, depending on how it was created.

**Comment by og900aero** [comment]
> Thanks for the verification.
> 

I thank you for your help :)

> > How can I compile it with this commit? Or just git clone again and make install?
> 
> Yes, you can just build it with the latest sources.
> 

I got it. Works fine. No error message, mounted right.

> > ```
> > -rw-r--r-- shyciii/users        599 2023-05-28 16:51 .gtkrc-2.0
> > hrw-r--r-- shyciii/users          0 2023-05-28 16:51 .gtkrc-2.0 link to .gtkrc-2.0
> > ```
> 
> Here it is. Your archive contains an entry which is a hard link... to itself.
> 
> I had never seen that before. Now that I know it can happen, I modified `fuse-archive` to simply ignore (i.e. skip) these redundant entries. Example:
> 
> ```
> fuse-archive --verbose /home/Data/Linux/Backup/valami.tar.zst mnt |& grep Skip
> ```
> 
> You should see the redundant link for `/.gtkrc-2.0` getting skipped.
> 
> Your archive might also contain other redundant links, depending on how it was created.

Me too. That is, it is interesting that it is a backup. There's another version of it 1-2 weeks earlier, and it doesn't have this error. If I rebuild the backup, it doesn't have this error either. So I don't know what could have happened to cause this error in a backup when I use the same command: tar -cavf /home/Data/Linux/Backup/home_backup_debian.tar.zst .config/

**Comment by fdegros** [comment]
All good.

> So I don't know what could have happened to cause this error in a backup when I use the same command: tar -cavf /home/Data/Linux/Backup/home_backup_debian.tar.zst .config/

This [`libarchive` bug report](https://github.com/libarchive/libarchive/issues/1381) describes a way to create such a redundant hard link in a TAR archive.

I released [`fuse-archive` 1.4](https://github.com/google/fuse-archive/releases/tag/v1.4) with the bug fix for this particular use case, which is to simply detect and skip the redundant hard links.
